### ISSUE TITLE
Optimize A4->A5 Rendering for Common Formats

### ISSUE THREAD

**Issue by NewCreature** [NewCreature/Allegro-Legacy#1]
The current method of using al_lock_bitmap() and al_put_pixel() is pretty slow. I was able to get a huge performance boost by modifying the locked data directly for 8-bit rendering. These optimizations need to be implemented for the other bit depths as well.

**Comment by ZoriaRPG** [comment]
AG4 used to have colour conversion in ASM. This was 200%+ faster on Windows, than the compiled output of ag44's libs using MSVC08. I'm unsure why, but it may be worthwhile investigating this.

**Comment by NewCreature** [comment]
I won't be supporting any assembly routines in Allegro Legacy. Any color conversions should work plenty fast on any modern hardware that will be running the code anyway.

Color conversions that happen inside Allegro 4's code aren't what this issue is about, ultimately. I am writing new routines that take color data stored in Allegro 4 bitmaps and converts it to whatever format the Allegro 5 bitmap needs.

If you need those assembly routines for an old piece of hardware, chances are you will want to be using Allegro 4 directly on that hardware instead of going through Allegro 5 with Allegro Legacy. It's dead simple to have your program support compiling with both Allegro 4 and Allegro Legacy by using `#ifdef` statements in the appropriate places.

**Comment by mlt** [comment]
Indeed, pixel by pixel copying is extremely slow. Wouldn't it be better if we could [decorate a4 bitmap](https://github.com/kazzmir/allegro4-to-5/blob/master/allegro4/include/gfx.h#L34) manipulations to [work on ALLEGRO_BITMAP directly](https://github.com/kazzmir/allegro4-to-5/blob/master/allegro4/allegro.c#L1050-L1088)? That would give hardware acceleration. That library is lacking many things already in AL but I think it has quite a potential.

**Comment by NewCreature** [comment]
> Wouldn't it be better if we could [decorate a4 bitmap](https://github.com/kazzmir/allegro4-to-5/blob/master/allegro4/include/gfx.h#L34) manipulations to [work on ALLEGRO_BITMAP directly](https://github.com/kazzmir/allegro4-to-5/blob/master/allegro4/allegro.c#L1050-L1088)?

That depends on your goals. This project is not about remapping all of Allegro 4's API onto Allegro 5's API. It is about providing a sort of platform-agnostic set of drivers for Allegro 4 that allows your Allegro 4 code to run on modern systems.

Many Allegro 4 programs were written with pixel-by-pixel manipulation in mind with no thought given to possible hardware acceleration. Mapping Allegro 4 calls onto Allegro 5 calls to get hardware acceleration comes at the cost of possibly inaccurately running the original software.

For example, you might draw lines and other primitives directly onto a `BITMAP`, which will use Allegro 4's software drawing algorithms. If you mapped that onto Allegro 5's primitives add-on, the output would likely differ due to the algorithms generating the output being different.

There are also things people used in Allegro 4 that you just can't do in Allegro 5. Did you ever utilize the `line` member of a `BITMAP`? I used it myself and saw it in other people's code as well. It was even encouraged by the manual.

With allegro4-to-5 you'll likely need to make some changes to your code to get it up and running. That doesn't make it a bad library. It just has different goals, or at least is taking a different approach.

I made Allegro Legacy because I have a lot of Allegro 4 code from past projects that I intend on supporting into the future. Ultimately, this project is about supporting older software. Allegro Legacy allows you to build your Allegro 4 programs without having to modify a single line of code.

To bring it back on topic, the A4->A5 rendering path is intended to provide a bridge between `BITMAP`s and `ALLEGRO_BITMAP`s, for handling final output onto the screen. As such, it is likely to only be done once per render frame, and thus should be fast enough for the kind of low resolution graphics that were typical of Allegro 4 programs.

For my purposes, the optimizations for 8-bit rendering were all I needed. I never ventured into using higher color depths when using Allegro 4, so I don't actually need to implement the other optimizations. I did intend to optimize some of the other format conversions, but never got around to it.

**Comment by mlt** [comment]
I don't recollect calling it "bad". If I did, I apologize. I do recognize different goals and tradeoffs associated with that.
Indeed, I saw `line` in a425 library as well as failed parts of examples (IIRC some flames). But I also have a patch for conditional compilation to remove it along with constant heap de-/allocation if not used explicitly.
I also recognize that it is not so straightforward to find substitutes for like `do_line` and a dashed line as it is better done with shaders.
I also suspect that some boost is possible with GLSL wrt AL and a5 rendering. Even for low-res and 8-bit, that would avoid double for-loops and expensive color conversion in favor of LUT in fragment/color shader.
